### PR TITLE
[hdf5] Update to 1.12.2

### PR DESCRIPTION
**Pull request by Thieum**
Updates hdf5 to 1.12.2 : https://github.com/HDFGroup/hdf5/releases/tag/hdf5-1_12_2

- #### What does your PR fix?
Updates hdf5 to 1.12.2

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
 no change (tested locally on windows), N/A

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
Yes

### CODE REVIEW

**Comment by JonLiu1993** [comment]
@Thieum ,Thanks for your pr, looks ci tested failed on many triplet
This is error log, please take a look:
````
CMake Error at /mnt/vcpkg-ci/installed/x64-linux/share/hdf5/vcpkg-cmake-wrapper.cmake:17 (_find_package):
  Could not find a package configuration file provided by "HDF5" with any of
  the following names:

    HDF5Config.cmake
    hdf5-config.cmake

  Add the installation prefix of "HDF5" to CMAKE_PREFIX_PATH or set
  "HDF5_DIR" to a directory containing one of the above files.  If "HDF5"
  provides a separate development package or SDK, be sure it has been
  installed.
````

File: ports/hdf5/portfile.cmake [hunk 1]
```diff
@@ -91,16 +91,16 @@ if(VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_TARGET_IS_MINGW AND VCPKG_LIBRARY_LINKA
...
 endif()
 
-file(READ "${CURRENT_PACKAGES_DIR}/share/hdf5/hdf5-config.cmake" contents)
+file(READ "${CURRENT_PACKAGES_DIR}/share/hdf5-config.cmake" contents)
 string(REPLACE [[${HDF5_PACKAGE_NAME}_TOOLS_DIR "${PACKAGE_PREFIX_DIR}/bin"]]
                [[${HDF5_PACKAGE_NAME}_TOOLS_DIR "${PACKAGE_PREFIX_DIR}/tools/hdf5"]]
                contents ${contents}
 )
-file(WRITE "${CURRENT_PACKAGES_DIR}/share/hdf5/hdf5-config.cmake" ${contents})
+file(WRITE "${CURRENT_PACKAGES_DIR}/share/hdf5-config.cmake" ${contents})
```
Comment (Neumann-A) [inline ports/hdf5/portfile.cmake:99]: this means the file gets installed/moved into the wrong directory.

**Comment by JonLiu1993** [comment]
@Thieum ,I  tested the features on my locally machine by command "./vcpkg install hdf5[*]:x86-windows" but failed,
There is error log, pleasae take a look:
````
CMake Error at config/cmake_ext_mod/HDFUseFortran.cmake:16 (enable_language):
  No CMAKE_Fortran_COMPILER could be found.

  Tell CMake where to find the compiler by setting either the environment
  variable "FC" or the CMake cache entry CMAKE_Fortran_COMPILER to the full
  path to the compiler, or to the compiler name if it is in the PATH.
Call Stack (most recent call first):
  CMakeLists.txt:1066 (include)

````

**Comment by Neumann-A** [comment]
@JonLiu1993 That is ok. (If you are only testing on windows)

**Comment by JonLiu1993** [comment]
> @JonLiu1993 That is ok. (If you are only testing on windows)

@Neumann-A, I re-tested the feature locally, I excluded the fortran feature, and the command _./vcpkg install hdf5[cpp,parallel,szip,threadsafe,tools,utils,zlib]_ still failed, this is the error log:
````
-- Fixing pkgconfig file: F:/Feature-test/tree-sitter/vcpkg/packages/hdf5_x86-windows/lib/pkgconfig/hdf5_cpp.pc
-- Fixing pkgconfig file: F:/Feature-test/tree-sitter/vcpkg/packages/hdf5_x86-windows/lib/pkgconfig/hdf5_hl.pc
-- Fixing pkgconfig file: F:/Feature-test/tree-sitter/vcpkg/packages/hdf5_x86-windows/lib/pkgconfig/hdf5_hl_cpp.pc
-- Using cached msys-mingw-w64-i686-pkg-config-0.29.2-3-any.pkg.tar.zst.
-- Using cached msys-mingw-w64-i686-libwinpthread-git-9.0.0.6373.5be8fcd83-1-any.pkg.tar.zst.
-- Using msys root at F:/Feature-test/tree-sitter/vcpkg/downloads/tools/msys2/9a1ec3f33446b195
-- Fixing pkgconfig file: F:/Feature-test/tree-sitter/vcpkg/packages/hdf5_x86-windows/debug/lib/pkgconfig/hdf5.pc
-- Fixing pkgconfig file: F:/Feature-test/tree-sitter/vcpkg/packages/hdf5_x86-windows/debug/lib/pkgconfig/hdf5_cpp.pc
-- Fixing pkgconfig file: F:/Feature-test/tree-sitter/vcpkg/packages/hdf5_x86-windows/debug/lib/pkgconfig/hdf5_hl.pc
-- Fixing pkgconfig file: F:/Feature-test/tree-sitter/vcpkg/packages/hdf5_x86-windows/debug/lib/pkgconfig/hdf5_hl_cpp.pc
CMake Error at scripts/cmake/vcpkg_copy_tools.cmake:70 (message):
  Couldn't find tool "h5perf_serial":

      "F:/Feature-test/vcpkg/packages/hdf5_x86-windows/bin/h5perf_serial.exe" does not exist
Call Stack (most recent call first):
  ports/hdf5/portfile.cmake:120 (vcpkg_copy_tools)
  scripts/ports.cmake:146 (include)


error: building hdf5:x86-windows failed with: BUILD_FAILED
Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
    https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+hdf5
You can submit a new issue at:
    https://github.com/microsoft/vcpkg/issues/new?template=report-package-build-failure.md&title=[hdf5]+Build+error
Include '[hdf5] Build error' in your bug report title, the following version information in your bug description, and attach any relevant failure logs from above.
    vcpkg-tool version: 2022-05-05-67e17c1782801cf481be9ac0b3765dff3e4bdeb8
    vcpkg-scripts version: 3d0cfc7a1 2022-05-18 (2 days ago)
````

**Comment by Thieum** [comment]
@JonLiu1993 added a guard for the optional tool.

File: ports/hdf5/portfile.cmake [hunk 2]
```diff
@@ -115,6 +115,11 @@ if(FEATURES MATCHES "tools")
         endif()
     endforeach()
 
+    if (EXISTS "${CURRENT_PACKAGES_DIR}/bin/h5perf_serial${VCPKG_TARGET_EXECUTABLE_SUFFIX}"
```
Comment (dg0yt) [inline ports/hdf5/portfile.cmake:118]: Do we know when it gets installed? `EXISTS` may sometimes hide unexpected changes.

File: ports/hdf5/portfile.cmake [hunk 3]
```diff
@@ -115,6 +115,11 @@ if(FEATURES MATCHES "tools")
...
+        AND EXISTS "${CURRENT_PACKAGES_DIR}/debug/bin/h5perf_serial${VCPKG_TARGET_EXECUTABLE_SUFFIX}")
```
Comment (dg0yt) [inline ports/hdf5/portfile.cmake:119]: This doesn't work for single-config builds (no `debug` variant).

**Review by dg0yt — CHANGES_REQUESTED** [review]

File: ports/hdf5/portfile.cmake (hunk 2 above)
Comment (Thieum) [reply ports/hdf5/portfile.cmake:118]: it gets installed on the build server, but not on local machines so far. I didn't pinpoint the difference yet.

File: ports/hdf5/portfile.cmake (hunk 3 above)
Comment (Thieum) [reply ports/hdf5/portfile.cmake:119]: I tried to reproduce the behavior from the other tools, I'm not familiar with single-config builds but wouldn't we have the same issue with this?
```
if (NOT EXISTS "${CURRENT_PACKAGES_DIR}/bin/${HDF5_TOOL}${VCPKG_TARGET_EXECUTABLE_SUFFIX}"
            OR NOT EXISTS "${CURRENT_PACKAGES_DIR}/debug/bin/${HDF5_TOOL}${VCPKG_TARGET_EXECUTABLE_SUFFIX}")
```

File: ports/hdf5/portfile.cmake (hunk 2 above)
Comment (dg0yt) [reply ports/hdf5/portfile.cmake:118]: > it gets installed on the build server, but not on local machines so far. I didn't pinpoint the difference yet.

That's why dislike the `EXISTS`: It indicates that we lack control over a particular build output. It is going to surprise users if they try to build the "same" configuration on another machine, but don't get an output that they might expect.

Comment (dg0yt) [reply ports/hdf5/portfile.cmake:118]: Looking at https://github.com/HDFGroup/hdf5/blob/hdf5-1_12_2/tools/src/h5perf/CMakeLists.txt:
It installs `${H5_DEP_EXECUTABLES}`.
But this variables isn't filled as a list, but overwritten by the last tool which is enabled. Order of tools:
1. h5perf_serial
2. if H5_HAVE_PARALLEL 
    - if UNIX: perf (always overwritten by the next one)
    - h5perf

I don't if this was intented upstream. But for now it should be enough to control the tool name with the `parallel` feature?

File: ports/hdf5/portfile.cmake [hunk 4]
```diff
@@ -116,6 +115,12 @@ if(FEATURES MATCHES "tools")
         endif()
     endforeach()
 
+    if(FEATURES MATCHES "parallel")
```
Comment (dg0yt) [inline ports/hdf5/portfile.cmake:118]: I see the bad example with `tools`, but I would prefer an exact test:
~~~cmake
if("parallel" IN_LIST FEATURES)
~~~

**Review by dg0yt — CHANGES_REQUESTED** [review]
Thanks for the update!

**Comment by Thieum** [comment]
@JonLiu1993 I think all the comments have been fixed. Let me know if anything else can be improved!

**Comment by JonLiu1993** [comment]
All features are tested successufully in the following triplet:
- x86-windows
- x64-windows
- x64-windows-static

**Review by ras0219-msft — APPROVED** [review]

**Comment by ras0219-msft** [comment]
Thanks for the update!

**Pull request merged**